src/style-spec/diff.js

```diff
@@ -111,25 +141,25 @@ function diffSources(before, after, commands, sourcesRemoved) {
     for (sourceId in before) {
         if (!before.hasOwnProperty(sourceId)) continue;
         if (!after.hasOwnProperty(sourceId)) {
-            commands.push({ command: operations.removeSource, args: [sourceId] });
-            sourcesRemoved[sourceId] = true;
+            removeSource(sourceId, commands, sourcesRemoved);
         }
     }
 
     // look for sources to add/update
     for (sourceId in after) {
         if (!after.hasOwnProperty(sourceId)) continue;
         if (!before.hasOwnProperty(sourceId)) {
-            commands.push({ command: operations.addSource, args: [sourceId, after[sourceId]] });
+            addSource(sourceId, after, commands);
         } else if (!isEqual(before[sourceId], after[sourceId])) {
             if (before[sourceId].type === 'geojson' && after[sourceId].type === 'geojson') {
-                // geojson sources use setGeoJSONSourceData command to update
-                commands.push({ command: operations.setGeoJSONSourceData, args: [sourceId, after[sourceId].data] });
+                if (canUpdateGeoJSON(before, after, sourceId)) {
+                    commands.push({ command: operations.setGeoJSONSourceData, args: [sourceId, after[sourceId].data] });
+                } else {
+                    updateSource(sourceId, after, commands, sourcesRemoved);
+                }
```

Comment:
Nit: you can move the `canUpdateGeoJSON` condition to the parent `if`, which will remove the need for the `else` above.

Reply:
good point will do

Reply:
changes made